[HDR] Opening a web page with HDR images on SDR screen then moving it to HDR screen does not show HDR contents.

mattwoodrow opened the pull request:
#### 8bc59df1bc504b8ee4eb1d1adb3c51f4ae4eaf8f
<pre>
[HDR] Opening a web page with HDR images on SDR screen then moving it to HDR screen does not show HDR contents.
<a href="https://bugs.webkit.org/show_bug.cgi?id=295443">https://bugs.webkit.org/show_bug.cgi?id=295443</a>
&lt;<a href="https://rdar.apple.com/154908419">rdar://154908419</a>&gt;

Reviewed by Mike Wyrzykowski.

If the screen changes to HDR, the value of Document::drawsHDRContent can change,
and we need to check compositing decisions again.

This should change the value passed to GraphicsLayer::setDrawsHDRContent, so we
no longer need to notify the GraphicsLayer tree of these changes.

Also fixes a bug where changing the pixel format of a RemoteLayerBackingStore
wasn&apos;t clearing the m_maxRequestedEDRHeadroom value, so this remained set on the
otherwise-SDR layer when moving to an SDR screen.

* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::screenSupportedContentsFormatsChanged): Deleted.
* Source/WebCore/page/LocalFrame.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::updateScreenSupportedContentsFormats):
* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(WebCore::GraphicsLayer::noteScreenSupportedContentsFormatsChangedIncludingDescendants): Deleted.
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::deviceOrPageScaleFactorChanged):
(WebCore::GraphicsLayer::screenSupportedContentsFormatsChanged): Deleted.
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::ensureStructuralLayer):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::screenSupportedContentsFormatsChanged): Deleted.
* Source/WebCore/rendering/RenderLayerCompositor.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::clearBackingStore):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.mm:
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::clearBackingStore):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.mm:
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::clearBackingStore):
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::ensureBackingStore):

Canonical link: <a href="https://commits.webkit.org/297042@main">https://commits.webkit.org/297042@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fb4821abe53dd553ef359d8116500354cfc695af

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110394 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30053 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20486 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116419 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60648 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112357 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30732 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38640 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83948 "Found 1 new test failure: imported/w3c/web-platform-tests/html/semantics/forms/constraints/infinite_backtracking.html (failure)") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113342 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24521 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99400 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64390 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23882 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17540 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60215 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93902 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17596 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119210 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37434 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27772 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92915 "Found 9 new test failures: imported/w3c/web-platform-tests/css/css-masking/clip-path/clip-path-descendant-text-mutated-001.html imported/w3c/web-platform-tests/css/css-text-decor/text-shadow/basic-opacity.html imported/w3c/web-platform-tests/css/css-text-decor/text-shadow/basic.html imported/w3c/web-platform-tests/css/css-text/hyphens/hyphens-vs-float-clearance-002.html imported/w3c/web-platform-tests/css/css-transforms/animation/transform-non-invertible-discrete-interpolation.html imported/w3c/web-platform-tests/css/css-view-transitions/massive-element-left-of-viewport-offscreen-new.html imported/w3c/web-platform-tests/css/css-view-transitions/massive-element-on-top-of-viewport-offscreen-new.html imported/w3c/web-platform-tests/css/css-view-transitions/pseudo-with-classes-exit.html imported/w3c/web-platform-tests/css/filter-effects/filters-drop-shadow-002.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37807 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95669 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92738 "Found 1 new API test failure: /WebKitGTK/TestWebKitPolicyClient:/webkit/WebKitPolicyClient/new-window-policy (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23621 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37736 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15466 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33377 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37329 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42800 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36991 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40331 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38699 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/page/Page.cpp:
```diff
@@ -1746,6 +1746,14 @@ void Page::updateScreenSupportedContentsFormats()
+            return;
+
+        if (RefPtr view = document.view())
+            view->setDescendantsNeedUpdateBackingAndHierarchyTraversal();
```
If this going to traverse all the layers in the view, do we still need to call above
```
for (auto& rootFrame : m_rootFrames)
    rootFrame->screenSupportedContentsFormatsChanged();
```
`GraphicsLayerCA::setDrawsHDRContent()` will trigger `DrawsHDRContentChanged` if `drawsHDRContent` changes.

mattwoodrow replied:
Indeed! Tested and confirmed that it works.

Will update with the old code removed.

mattwoodrow committed:
[HDR] Opening a web page with HDR images on SDR screen then moving it to HDR screen does not show HDR contents.
https://bugs.webkit.org/show_bug.cgi?id=295443
<rdar://154908419>

Reviewed by Mike Wyrzykowski.

If the screen changes to HDR, the value of Document::drawsHDRContent can change,
and we need to check compositing decisions again.

This should change the value passed to GraphicsLayer::setDrawsHDRContent, so we
no longer need to notify the GraphicsLayer tree of these changes.

Also fixes a bug where changing the pixel format of a RemoteLayerBackingStore
wasn't clearing the m_maxRequestedEDRHeadroom value, so this remained set on the
otherwise-SDR layer when moving to an SDR screen.

* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::screenSupportedContentsFormatsChanged): Deleted.
* Source/WebCore/page/LocalFrame.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::updateScreenSupportedContentsFormats):
* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(WebCore::GraphicsLayer::noteScreenSupportedContentsFormatsChangedIncludingDescendants): Deleted.
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::deviceOrPageScaleFactorChanged):
(WebCore::GraphicsLayer::screenSupportedContentsFormatsChanged): Deleted.
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::ensureStructuralLayer):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::screenSupportedContentsFormatsChanged): Deleted.
* Source/WebCore/rendering/RenderLayerCompositor.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::clearBackingStore):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.mm:
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::clearBackingStore):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.mm:
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::clearBackingStore):
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::ensureBackingStore):

Canonical link: https://commits.webkit.org/297042@main
